### PR TITLE
[G-API]: Add GArray initialization support

### PR DESCRIPTION
**Pull request by mpashchenkov**
## Add GArray initialization support

Added support initialization for GArray. CONST_VALUE used for this.
You can create an empty or filled std::vector and put or move it into a GArray.

Example:

```
std::vector<cv::Mat> vec; // or vec = {some values}
cv::GComputation pp([&]() {
    cv::GMat in;
    cv::GArray<cv::GMat> garr(vec); // or garr(std::move(vec))
    return cv::GComputation(custom::SomeOp::on(in, garr));
});
```
Possible:
- `cv::GArray<cv::GMat> garr(vec)`;
- `cv::GArray<cv::GMat> garr(rval)`;
- `cv::GArray<cv::GMat> garr({some values})`.

Impossible:

- `cv::GArray<cv::GMat> garr = vec; // explicit`;
- `cv::GArray<cv::GMat> garr = std::vector<cv::Mat>(some size); // explicit`;
- `cv::GArray<cv::GMat> garr = std::vector<cv::Mat>()`.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### Magic commands

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2020.3.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by mpashchenkov** [comment]
@AsyaPronina Please review it, when build is ready and green)

File: modules/gapi/include/opencv2/gapi/garray.hpp [hunk 1]
```diff
@@ -336,17 +338,21 @@ namespace detail
 template<typename T> class GArray
 {
 public:
+    // Host type (or Flat type) - the type this GArray is actually
+    // specified to.
+    using HT = typename detail::flatten_g<typename std::decay<T>::type>::type;
```
Comment (OrestChura) [inline modules/gapi/include/opencv2/gapi/garray.hpp:343]: is it correct to move `HT` from `private` to `public`? Were there any reasons to put it into `private` originally?

File: modules/gapi/test/gapi_array_tests.cpp [hunk 2]
```diff
@@ -164,6 +177,25 @@ TEST(GArray, TestIntermediateOutput)
     EXPECT_EQ(10,  out_count[0]);
 }
 
+TEST(GArray, GArrayInitialization)
```
Comment (OrestChura) [inline modules/gapi/test/gapi_array_tests.cpp:180]: Mb name the test case more specifically? Like, `GArrayConstValueInitialization`
Who knows, perhaps there'll be other initializations in the future...

**Review by OrestChura — COMMENTED** [review]
Hi! Looks good to me :)

File: modules/gapi/include/opencv2/gapi/garray.hpp (hunk 1 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/garray.hpp:343]: We need HT in the public constructor. Put private zone under public zone - bad behavior.

File: modules/gapi/test/gapi_array_tests.cpp (hunk 2 above)
Comment (mpashchenkov) [reply modules/gapi/test/gapi_array_tests.cpp:180]: M B

File: modules/gapi/include/opencv2/gapi/garray.hpp (hunk 1 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/garray.hpp:343]: I assume you meant "put private above"

Yeah, agree that swapping them isn't a good idea

Anyway, it doesn't seem to cause any severe problems..

OK

File: modules/gapi/test/gapi_array_tests.cpp [hunk 3]
```diff
@@ -28,6 +28,10 @@ G_TYPED_KERNEL(CountCorners,   <GScalar(GPointArray)>,  "test.array.in")
 {
     static GScalarDesc outMeta(const GArrayDesc &) { return empty_scalar_desc(); }
 };
+G_TYPED_KERNEL(PointIncrement, <GPointArray(GMat, GPointArray)>, "test.array.initialization")
```
Comment (AsyaPronina) [inline modules/gapi/test/gapi_array_tests.cpp:31]: May be better to define operation ID as "point_increment" ?

**Comment by AsyaPronina** [comment]
I think that case with GArray const value initialization along with these impossible cases:

cv::GArray<cv::GMat> garr = vec; // explicit;
cv::GArray<cv::GMat> garr = std::vector<cv::Mat>(some size); // explicit;
cv::GArray<cv::GMat> garr = std::vector<cv::Mat>().
Impossible under lambda:

cv::GArray<cv::GMat> garr(std::vector<cv::Mat>(some size)); // memory error.

should be described in the code somewhere, for example, in GArray class documentation.

Why the last case is impossible?

File: modules/gapi/src/api/gorigin.cpp [hunk 4]
```diff
@@ -21,8 +21,14 @@ cv::GOrigin::GOrigin(GShape s,
 }
 
 cv::GOrigin::GOrigin(GShape s, cv::gimpl::ConstVal v)
-    : shape(s), node(cv::GNode::Const()), value(v), port(INVALID_PORT), kind(cv::detail::OpaqueKind::CV_UNKNOWN)
+    : shape(s), node(cv::GNode::Const()), value(v), port(INVALID_PORT)
```
Comment (mpashchenkov) [inline modules/gapi/src/api/gorigin.cpp:24]: @smirnov-alexey about kind for vectorRef

File: modules/gapi/test/gapi_array_tests.cpp (hunk 3 above)
Comment (mpashchenkov) [reply modules/gapi/test/gapi_array_tests.cpp]: Fixed

**Review by OrestChura — COMMENTED** [review]
Everything is wonderful!

File: modules/gapi/src/api/gorigin.cpp [hunk 5]
```diff
@@ -21,8 +21,14 @@ cv::GOrigin::GOrigin(GShape s,
...
 {
+    switch (v.index()) {
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/src/api/gorigin.cpp]: Fixed

**Comment by smirnov-alexey** [comment]
@mpashchenkov please, fill the PR readiness checklist

File: modules/gapi/include/opencv2/gapi/garray.hpp [hunk 6]
```diff
@@ -328,17 +329,21 @@ namespace detail
 template<typename T> class GArray
 {
 public:
+    // Host type (or Flat type) - the type this GArray is actually
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/garray.hpp:332]: Why are these changes here? Don't they suppose to be already in master?

File: modules/gapi/src/api/gorigin.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply modules/gapi/src/api/gorigin.cpp:24]: Looks good, thanks you

**Comment by smirnov-alexey** [comment]
Overall looks good

File: modules/gapi/include/opencv2/gapi/garray.hpp (hunk 6 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/garray.hpp:332]: We need HT in the public constructor. Put private zone above public zone - bad behavior.

**Review by OrestChura — APPROVED** [review]

File: modules/gapi/include/opencv2/gapi/garray.hpp (hunk 6 above)
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/garray.hpp:332]: I guess it's fine to have HT in public since it's almost the original T which is known

File: modules/gapi/include/opencv2/gapi/garray.hpp [hunk 7]
```diff
@@ -328,17 +329,21 @@ namespace detail
...
+    // specified to.
+    using HT = typename detail::flatten_g<typename std::decay<T>::type>::type;
+
+    explicit GArray(const std::vector<HT>& v) // Constant value constructor
+        : m_ref(detail::GArrayU(detail::VectorRef(v))) { putDetails(); }
+    explicit GArray(std::vector<HT>&& v)      // Constant value move-constructor
```
Comment (AsyaPronina) [inline modules/gapi/include/opencv2/gapi/garray.hpp:338]: It is not constant value move constructor, it is just a move constructor

Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/garray.hpp:338]: I think constant value may be misinterpreted here and should be described in more details with mentioned context or just dropped

File: modules/gapi/test/gapi_array_tests.cpp [hunk 8]
```diff
@@ -28,6 +28,10 @@ G_TYPED_KERNEL(CountCorners,   <GScalar(GPointArray)>,  "test.array.in")
 {
     static GScalarDesc outMeta(const GArrayDesc &) { return empty_scalar_desc(); }
 };
+G_TYPED_KERNEL(PointIncrement, <GPointArray(GMat, GPointArray)>, "test.point_increment")
+{
+    static GArrayDesc outMeta(const GMatDesc&, const GArrayDesc &) { return empty_array_desc(); }
```
Comment (AsyaPronina) [inline modules/gapi/test/gapi_array_tests.cpp:33]: Please remove space before second ``&`` or add before first ``&``

File: modules/gapi/test/gapi_array_tests.cpp [hunk 9]
```diff
@@ -57,6 +61,15 @@ GAPI_OCV_KERNEL(OCVCountCorners, ThisTest::CountCorners)
     }
 };
 
+GAPI_OCV_KERNEL(OCVPointIncrement, ThisTest::PointIncrement)
+{
+    static void run(cv::Mat, const std::vector<cv::Point> &in, std::vector<cv::Point> &out)
```
Comment (AsyaPronina) [inline modules/gapi/test/gapi_array_tests.cpp:66]: ``const cv::Mat&``

File: modules/gapi/test/gapi_array_tests.cpp [hunk 10]
```diff
@@ -164,6 +177,44 @@ TEST(GArray, TestIntermediateOutput)
...
+TEST(GArray, GArrayConstRValInitialization)
+{
+    std::vector<cv::Point> ref_vec {Point(1,1), Point(2,2), Point(3,3)};
+    std::vector<cv::Point> out_vec;
+    cv::Mat in_mat;
+
+    cv::GComputationT<ThisTest::GPointArray(cv::GMat)> c([&](cv::GMat in)
+    {
+        ThisTest::GPointArray test_garray(
+            { Point(0,0), Point(1,1), Point(2,2) } ); // Rvalue initialization
```
Comment (AsyaPronina) [inline modules/gapi/test/gapi_array_tests.cpp:208]: I think it is safely can be placed on the previous line

File: modules/gapi/include/opencv2/gapi/garray.hpp (hunk 7 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/garray.hpp]: Fixed (dropped).

File: modules/gapi/test/gapi_array_tests.cpp (hunk 8 above)
Comment (mpashchenkov) [reply modules/gapi/test/gapi_array_tests.cpp]: Fixed.

File: modules/gapi/test/gapi_array_tests.cpp (hunk 10 above)
Comment (mpashchenkov) [reply modules/gapi/test/gapi_array_tests.cpp]: Fixed.

File: modules/gapi/test/gapi_array_tests.cpp (hunk 9 above)
Comment (mpashchenkov) [reply modules/gapi/test/gapi_array_tests.cpp]: Fixed.

**Comment by mpashchenkov** [comment]
@dmatveev, please review (the last word).

File: modules/gapi/src/api/gorigin.cpp [hunk 11]
```diff
@@ -21,8 +21,15 @@ cv::GOrigin::GOrigin(GShape s,
...
-    : shape(s), node(cv::GNode::Const()), value(v), port(INVALID_PORT), kind(cv::detail::OpaqueKind::CV_UNKNOWN)
+    : shape(s), node(cv::GNode::Const()), value(v), port(INVALID_PORT)
 {
+    switch (v.index())
+    {
+    case cv::gimpl::ConstVal::index_of<detail::VectorRef>():
+        kind = util::get<detail::VectorRef>(v).getKind();
+    default:
+        kind = cv::detail::OpaqueKind::CV_UNKNOWN;
+    }
```
Comment (dmatveev) [inline modules/gapi/src/api/gorigin.cpp:32]:
```suggestion
kind = holds_alternative<VectorRef>(v)
    ? util::get<detail::VectorRef>(v).getKind()
    : cv::detail::OpaqueKind::CV_UNKNOWN;
```

**Review by dmatveev — APPROVED** [review]
Overall looks good!

File: modules/gapi/src/api/gorigin.cpp (hunk 11 above)
Comment (mpashchenkov) [reply modules/gapi/src/api/gorigin.cpp:32]: Ok. But, what if we want to add initialization for gopaque?

Comment (mpashchenkov) [reply modules/gapi/src/api/gorigin.cpp]: Fixed.

Comment (dmatveev) [reply modules/gapi/src/api/gorigin.cpp]: Good point btw, extending it to opaque will really look better with `switch()`. You could leave a placeholder here before `default` to indicate that `OpaqueRef` will be the next case - then it'd be more clear

**Comment by dmatveev** [comment]
👍 go ahead with merge. Thanks!
